docs/proposals/20210524-managed-external-etcd.md

```diff
@@ -0,0 +1,570 @@
+
+##### Etcdadm bootstrap provider
+
+The Etcdadm bootstrap provider will convert a Machine to an etcd member by running `etcdadm` commands on it. When using the `etcdadm` CLI, the two main steps in etcd cluster creation are:
+- Choosing any one of the machines and running `etcdadm init` on it to create a single node etcd cluster.
+- Running `etcdadm join <first member's endpoint>` on the remaining machines to add them to the cluster created in the first step. This requires the CA certs to be present at an expected path on the machines.
+
+This bootstrap provider will perform the steps specified above using a custom controller that processes the EtcdadmConfig CRs. This controller is similar in design to the Kubeadm bootstrap provider controller. It follows the contract defined for [bootstrap providers](https://cluster-api.sigs.k8s.io/developer/providers/bootstrap.html#behavior). It performs the following actions upon receiving an EtcdadmConfig CR from the Workqueue:
+  - Each EtcdadmConfig object has a 1:1 mapping to the Machine object. The controller will get the owning Machine for the EtcdadmConfig object being processed. (The next section explains how the [Etcdadm cluster controller](#etcdadm-cluster-controller)) creates this association between an EtcdadmConfig and a Machine object).
+  - The controller then gets the Cluster object from the Machine.
+    - If the Cluster does not have the `ManagedExternalEtcdInitialized` condition set to true, the controller will choose this Machine to run the `etcdadm init` command.
+    - If the etcd cluster is initialized, meaning the `ManagedExternalEtcdInitialized` condition is true, the controller will make this Machine run the `etcdadm join` command.
+  - For each Machine, the controller will lookup a Secret by the name `{Cluster.Name}-managed-etcd` to get the CA cert-key for the external etcd cluster. (The next section explains how the etcdadm cluster controller generates this Secret).
+  - To add the certs to the required path on the Machines and run the required etcdadm commands, the controller will generate [cloud-init](https://cloudinit.readthedocs.io/) data for each member containing:
+    - A [`write_files` section](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#write-files) to write the CA cert-key pair at `etc/etcd/pki` path on each Machine.
+    - A [`runcmd` section](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#runcmd) containing the etcdadm commands, along with any user specified commands.
+  The controller then saves this cloud-init script as a Secret, and this Secret's name on the EtcdadmConfig.Status.DataSecretName field.
+  - Since the EtcdadmConfig.Status.DataSecretName field gets set as per the [bootstrap provider specifications](https://cluster-api.sigs.k8s.io/developer/providers/bootstrap.html), the infrastructure providers will use the data from this Secret to initiate the Machines with the cloud-init script.
+  
+##### Etcdadm cluster controller
```

Comment:
I think it will be useful to the cluster deprovisioning workflow, given that I assume some kind of syncronization is required there as well in order to ensure things are deleted in the right order